File: sdk/javascript-enhanced/src/crash/FlagentCrashReporter.ts
/**
 * Firebase Crashlytics-level crash reporting for Flagent.
 * Captures window.onerror and unhandledrejection, sends to Flagent backend.
 */

export interface FlagentCrashReporterConfig {
  /** Base URL of Flagent server (e.g. "http://localhost:18000") */
  baseUrl: string;
  /** Optional X-API-Key for tenant-scoped crash reports */
  apiKey?: string;
  /** Platform identifier, default "web" */
  platform?: string;
  /** Optional app version string */
  appVersion?: string;
}

interface CrashPayload {
  stackTrace: string;
  message: string;
  platform: string;
  appVersion?: string;
  deviceInfo?: string;
  breadcrumbs?: string | null;
  customKeys?: string | null;
  timestamp: number;
}

export class FlagentCrashReporter {
  private readonly baseUrl: string;
  private readonly apiKey?: string;
  private readonly platform: string;
  private readonly appVersion?: string;
  private onErrorHandler?: (event: ErrorEvent) => void;
  private onRejectionHandler?: (event: PromiseRejectionEvent) => void;

  constructor(config: FlagentCrashReporterConfig) {
    this.baseUrl = config.baseUrl.replace(/\/$/, '');
    this.apiKey = config.apiKey;
    this.platform = config.platform ?? 'web';
    this.appVersion = config.appVersion;
  }

  /** Install crash reporter. Call early in app startup. */
  install(): void {
    this.onErrorHandler = (event: ErrorEvent) => {
      const message = event.message ?? String(event.error ?? 'Unknown error');
      const stackTrace = event.error?.stack ?? `${event.filename}:${event.lineno}:${event.colno}`;
      this.send({ stackTrace, message });
    };
    this.onRejectionHandler = (event: PromiseRejectionEvent) => {
      const err = event.reason;
      const message = err?.message ?? String(err ?? 'Unhandled rejection');
      const stackTrace = err?.stack ?? 'No stack trace';
      this.send({ stackTrace, message });
    };
    if (typeof window !== 'undefined') {
      window.addEventListener('error', this.onErrorHandler);
      window.addEventListener('unhandledrejection', this.onRejectionHandler);
    }
  }

  /** Uninstall crash reporter. */
  uninstall(): void {
    if (typeof window !== 'undefined') {
      if (this.onErrorHandler) window.removeEventListener('error', this.onErrorHandler);
      if (this.onRejectionHandler) window.removeEventListener('unhandledrejection', this.onRejectionHandler);
    }
  }

  private send(payload: Partial<CrashPayload>): void {
    const full: CrashPayload = {
      stackTrace: payload.stackTrace ?? 'No stack trace',
      message: payload.message ?? 'Unknown error',
      platform: this.platform,
      appVersion: this.appVersion,
      deviceInfo: typeof navigator !== 'undefined' ? navigator.userAgent : undefined,
      breadcrumbs: null,
      customKeys: null,
      timestamp: Date.now(),
      ...payload,
    };
    const headers: Record<string, string> = { 'Content-Type': 'application/json' };
    if (this.apiKey) headers['X-API-Key'] = this.apiKey;
    fetch(`${this.baseUrl}/api/v1/crashes`, {
      method: 'POST',
      headers,
      body: JSON.stringify(full),
      keepalive: true,
    }).catch(() => {});
  }
}

File: sdk/kotlin-enhanced/src/jvmMain/kotlin/com/flagent/enhanced/crash/FlagentCrashReporter.kt
package com.flagent.enhanced.crash

import io.ktor.client.*
import io.ktor.client.engine.cio.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Crash reporter for Flagent (Firebase Crashlytics-level).
 * Installs UncaughtExceptionHandler to capture crashes and POST to Flagent.
 *
 * @param baseUrl Base URL of Flagent server (e.g. "http://localhost:18000")
 * @param apiKey Optional X-API-Key for tenant-scoped crash reports
 * @param platform Platform identifier: "android", "ios", "jvm"
 * @param appVersion Optional app version string
 * @param deviceInfo Optional device info (e.g. "Android 14, Pixel 7")
 */
class FlagentCrashReporter(
    private val baseUrl: String,
    private val apiKey: String? = null,
    private val platform: String = "jvm",
    private val appVersion: String? = null,
    private val deviceInfo: String? = null,
    httpClient: HttpClient? = null
) {
    private val client = httpClient ?: HttpClient(CIO) {
        install(ContentNegotiation) {
            json(Json { ignoreUnknownKeys = true })
        }
    }

    private val defaultHandler = Thread.getDefaultUncaughtExceptionHandler()

    private val crashHandler = Thread.UncaughtExceptionHandler { thread, throwable ->
        try {
            val stackTrace = throwable.stackTraceToString()
            val message = throwable.message ?: throwable.javaClass.simpleName
            val payload = CrashReportRequest(
                stackTrace = stackTrace,
                message = message,
                platform = platform,
                appVersion = appVersion,
                deviceInfo = deviceInfo,
                breadcrumbs = null,
                customKeys = null,
                timestamp = System.currentTimeMillis()
            )
            runBlocking {
                withContext(Dispatchers.IO) {
                    try {
                        client.post("$baseUrl/api/v1/crashes") {
                            contentType(ContentType.Application.Json)
                            apiKey?.let { header("X-API-Key", it) }
                            setBody(Json.encodeToString(CrashReportRequest.serializer(), payload))
                        }
                    } catch (_: Exception) { /* best-effort */ }
                }
            }
        } catch (_: Exception) { /* avoid secondary crash */ }
        defaultHandler?.uncaughtException(thread, throwable)
    }

    /**
     * Install crash reporter. Call early in app startup (e.g. Application.onCreate).
     */
    fun install() {
        Thread.setDefaultUncaughtExceptionHandler(crashHandler)
    }

    /**
     * Uninstall and restore previous handler.
     */
    fun uninstall() {
        Thread.setDefaultUncaughtExceptionHandler(defaultHandler)
    }

    @Serializable
    private data class CrashReportRequest(
        val stackTrace: String,
        val message: String,
        val platform: String,
        val appVersion: String? = null,
        val deviceInfo: String? = null,
        val breadcrumbs: String? = null,
        val customKeys: String? = null,
        val timestamp: Long
    )
}
